agent/lib/kontena/observable.rb

```diff
@@ -1,75 +1,201 @@
+          observable_registry.register(observable, self.current_actor)
+
+          self.links << observable_registry # registry monitors owner
+        end
+      end
+    end
+
+    # @param subject [Object] used to identify the Observable for logging purposes
+    def initialize(subject = nil)
+      @subject = subject
+      @mutex = Thread::Mutex.new
+      @observers = {}
+      @value = nil
+
+      # include the subject (owning actor class, other resource) in log messages
+      @logging_prefix = "#{self}"
+    end
+
+    # @return [String]
+    def to_s
+      "#{self.class.name}<#{@subject}>"
+    end
+
+    # @return [Object, nil] last updated value, or nil if not observable?
+    def get
+      @value
+    end
+
+    # Observable has updated, and has not reset. It might be crashed?
+    #
+    # @return [Boolean]
+    def observable?
+      !!@value
+    end
+
+    # Observable has an exception set.
+    #
+    # Calls to `add_observer` will raise.
+    #
+    def crashed?
+      Exception === @value
     end
 
-    # Registered Observers
+    # Observable has observers.
+    #
+    # NOTE: dead observers will only get cleaned out on the next update
     #
-    # @return [Hash{Observe => Celluloid::Proxy::Cell<Observer>}]
-    def observers
-      @observers ||= {}
+    # @return [Boolean]
+    def observed?
+      !@observers.empty?
     end
 
-    # The Observable has a value. Propagate it to any observing Actors.
+    # The Observable has a value. Propagate it to any observers.
     #
-    # This will notify any Observers, causing them to yield if ready.
+    # This will notify any Observers, causing them to yield/return if ready.
     #
-    # The value must be safe for access by multiple threads, even after this update,
-    # and even after any later updates. Ideally, it should be immutable (frozen).
+    # The value must be immutable and threadsafe: it must remain valid for use by other threads
+    # both after this update, and after any other future updates. Do not send a mutable object
+    # that gets invalidated in between updates.
+    #
+    # TODO: automatically freeze the value?
     #
     # @param value [Object]
+    # @raise [RuntimeError] Observable crashed
     # @raise [ArgumentError] Update with nil value
-    def update_observable(value)
+    def update(value)
+      raise RuntimeError, "Observable crashed: #{@value}" if crashed?
       raise ArgumentError, "Update with nil value" if value.nil?
-      debug "update: #{value}"
 
-      @observable_value = value
+      debug { "update: #{value}" }
 
-      notify_observers
+      set_and_notify(value)
     end
 
-    # The Observable no longer has a value
-    # This will notify any Observers, causing them to block yields until we update again
-    def reset_observable
-      @observable_value = nil
+    # Reset the observable value back into the initialized state.
+    # This will notify any Observers, causing them to block until we update again.
```

Comment:
Should probably avoid the use of the term `block` for this, and use something like `wait` instead. The observer debug logging used to say `observe blocked: Kontena::Observable<...>?` if it was waiting for something to get a value, but I already changed that to log `wait: ...`

If you do a sync `observe` after resetting the observable, then it will wait for the observable to get updated again.

Reply:
Yeah, the word `block` made this really hard to understand :D So probably good idea to switch `block` to something else.

Reply:
No more "block" terms now.